Trac 59165: Add Font Face and its tests

hellofromtonya opened the pull request:
Adds the new Font Face, a server-side `@font-face` styles generator and printer. Code is copied from Gutenberg.

Trac ticket: https://core.trac.wordpress.org/ticket/59165

## How to Test

Scenario 1 uses the Twenty Twenty-Three (TT3) theme. [Testing instructions are provided here](https://core.trac.wordpress.org/ticket/59165#comment:5) in the Trac ticket.

Scenario 2 is a classic site with a plugin that passes an array of fonts when invoking `wp_print_font_faces()`. [Testing instructions are provided here in the Trac ticket](https://core.trac.wordpress.org/ticket/59165#comment:7).

## For Code Reviewers

Font Face was developed in Gutenberg. The vast majority of its code comes from well-established font processing developed in multiple iterations of the Webfonts, Web Fonts, and Fonts API. You can find:

* the source code here in Gutenberg's `lib/compat/wordpress-6.4/fonts/` directory https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.4/fonts.
* PHPUnit tests live here https://github.com/WordPress/gutenberg/tree/trunk/phpunit/tests/fonts/font-face.

Changes made in this PR will be copied / merged back into Gutenberg.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

hellofromtonya committed:
Add Font Face and its tests

hellofromtonya committed:
Deprecates stopgap code

hellofromtonya commented:
Need to merge https://github.com/WordPress/gutenberg/pull/53805 to use `wp_get_global_settings()` instead of the private API. Commit will be pushed shortly.

Update: Done ✅  via https://github.com/WordPress/wordpress-develop/pull/5051/commits/e1da7ab4b169afab897403e6cc49e2840c26d29c.

Props @oandregal.

hellofromtonya committed:
Use wp_get_global_settings() instead of private API

Changes the resolver to use wp_get_global_settings() instead of
the chained objects to get the merged data layer.

Reference:
https://github.com/WordPress/gutenberg/pull/53805

Props @oandregal.

hellofromtonya committed:
Details fonts array structure

hellofromtonya commented on src/wp-includes/fonts/fonts.php:
```diff
@@ -0,0 +1,63 @@
+		null === $wp_font_face ||
+
+		// Ignore cache when automated test suites are running.
+		( defined( 'WP_RUN_CORE_TESTS' ) && WP_RUN_CORE_TESTS )
```
The intent:

* Use a new instance of `WP_Font_Face` for each PHPUnit test. Why? To avoid test-to-test interactions by ensuring a new object is used for each test.

* Else, it acts as a singleton whether in `WP_DEBUG`, prod, or development modes.

costdev replied:
Do the tests pass with a combination of `@runInSeparateProcess` and `@preserveGlobalState disabled` in the test docblocks?

If so, while it would slow down the tests, it removes the need for this test-related source code and separates implemention from the test suite. It may be worth considering if it hasn't been already.

hellofromtonya replied:
All of the Font Face tests would require running in their own separate processes. What would the test performance impact be?

* The code now: 00:05.447
* Running in separate processes: 00:36.257

That's fairly significant for 20 tests. Ouch.

hellofromtonya replied:
Whoopsie, I was wrong. Only the global function tests (i.e. for `wp_print_font_faces()`) need to run in a separate process, not all of the tests.

Re-running the `fonts` group tests with the `wp_print_font_faces()` tests in separate processes: 00:16.807

Adds about 11 seconds with the current tests, i.e. more unhappy tests are coming.

What do you think @costdev?

hellofromtonya replied:
Colin and I chatted in Slack. Commit https://github.com/WordPress/wordpress-develop/pull/5051 runs the tests in separate processes. 

Results: adds ~11 seconds to the run time of the test suites 😢 , but removes the test constant from the source code 😄 .

hellofromtonya commented on src/wp-includes/fonts/font-face/class-wp-font-face-resolver.php:
```diff
@@ -0,0 +1,154 @@
+	 * @return array Returns the font-families, each with their font-face variations.
+	 */
+	public static function get_fonts_from_theme_json() {
+		$settings = wp_get_global_settings();
```
@oandregal I merged in https://github.com/WordPress/gutenberg/pull/53805 and updated to use Core's `wp_get_global_settings()`.

felixarntz commented on src/wp-includes/fonts/font-face/class-wp-font-face.php:
```diff
@@ -0,0 +1,439 @@
+		 *     @type string $font-display The font-display property. Default 'fallback'.
+		 * }
+		 */
+		$this->font_face_property_defaults = apply_filters( 'wp_font_face_property_defaults', $this->font_face_property_defaults );
```
What is the rationale for introducing this filter? Unless we have a good use-case for it, I would advise to remove it. We can always add filters later if it becomes necessary, but we can't remove them once introduced. Any API extension point is a potential future challenge for backward compatibility breaks.

Additionally, specifically for the logic around `theme.json`, the existing filters have made previous attempts to cache the data unsuccessful as caching would not be able to maintain the filter logic reliably. So I'd prefer not to introduce more filters in this area unless absolutely necessary.

hellofromtonya replied:
Very good points. It was added to allow extenders to change the font-face properties and defaults, which are used in the validation process. But maybe that is not needed, especially at the point of introduction. The current list aligns to mdn web docs for `@font-face` properties https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face.

Thus, I agree with you and am removing it.

hellofromtonya replied:
Filter removed in commit https://github.com/WordPress/wordpress-develop/pull/5051/commits/4db7d421a6304ff31b0186d8247237a698c32073 ✅

felixarntz commented on src/wp-includes/fonts/font-face/class-wp-font-face.php:
```diff
@@ -0,0 +1,439 @@
+	 *             @type string          $font_style              Optional. The font-style property. Default 'normal'.
+	 *             @type string          $font-weight             Optional. The font-weight property. Default '400'.
+	 *             @type string          $font-display            Optional. The font-display property. Default 'fallback'.
+	 *             The following font-face properties are optional and will not be generated if not provided.
```
This line should be removed as it's not following the docs standards, there shouldn't be "general" explanations between `@type` annotations. Additionally, this explanation is confusing in the first place, as there are already some fields marked as "Optional." above it.

Let's simply rely on having the "Optional." annotation on each field as needed, which is already present.

hellofromtonya replied:
Removed in `wp_print_font_faces() in commit  https://github.com/WordPress/wordpress-develop/commit/31a3508a839cfa0427c2c471bc8513b0693e8ed9 ✅ 

And this reference was removed in `WP_Font_Face()` using your suggestion (thank you!) via commit https://github.com/WordPress/wordpress-develop/pull/5051/commits/f6c42668cf426456a5d5e36e1ad5d15636dc523e  ✅

felixarntz commented on src/wp-includes/fonts/font-face/class-wp-font-face.php:
```diff
@@ -0,0 +1,439 @@
+
+		// Check the font-family.
+		if ( empty( $font_face['font-family'] ) || ! is_string( $font_face['font-family'] ) ) {
+			trigger_error( 'Font font-family must be a non-empty string.' );
```
Why is this class using `trigger_error()` (here and below)? Core should always use the appropriate wrapper functions, such as `_doing_it_wrong()`, so that e.g. `WP_DEBUG` is respected.

hellofromtonya replied:
This change comes from @azaozz, who advise only using `_doing_it_wrong()` for the most severe cases. Since then, `trigger_error()` has been used for things like function/method parameter (input) validation. It exists in Core for things like PHP compat changes.

An architectural proposal for handling input validation (i.e. function/method parameter checks) is in the works which will also include new error handling.

Also noting the code being replaced also used `trigger_error()` for these same validation checks.

hellofromtonya replied:
Note: Here's a Trac ticket that proposes introducing `wp_trigger_error()` to compliment `_doing_it_wrong()` https://core.trac.wordpress.org/ticket/57686. @azaozz shares reasoning of why not use `_doing_it_wrong()` for these purposes:

>The _doing_it_wrong() function was introduced 12 years ago .. for "telling people the code they have written won't work"...
>
>However that convenience has resulted in shifting the initial purpose, which was to tell developers they are doing something wrong. Lately `_doing_it_wrong()` seems to be used in any context, not specifically to "tell off" the developers and get them to fix their code.

felixarntz replied:
@hellofromtonya Thank you for the additional context. I may be on board with a new function like `wp_trigger_error()`, but since that is not in place yet, I think the usage here should rather rely on `_doing_it_wrong()`, since in those cases it literally is the developer "doing it wrong" rather than something else leading to the problem.

As you mention, `trigger_error()` is currently mostly used in Core for things that deal with PHP compat, which is not the case here, and calling `trigger_error()` ignores `WP_DEBUG` etc constants which is the main reason I object to this change.

I think we should use `_doing_it_wrong()`, and then later we can update it to use `wp_trigger_error()` once that function has been added to Core (potentially immediately as part of https://core.trac.wordpress.org/ticket/57686).

azaozz replied:
> calling trigger_error() ignores WP_DEBUG

Yea, lets just add `wp_trigger_error()`. I know there are ideas to expand/extend its use in the future, but seems is it needed now :)

hellofromtonya replied:
>Yea, lets just add wp_trigger_error(). I know there are ideas to expand/extend its use in the future, but seems is it needed now :)

Sorry for my delay in responding (been sick). 

Thanks @felixarntz and @azaozz. I think you both make good points.

- I moved [Trac 57686](https://core.trac.wordpress.org/ticket/57686#comment:7) back into the 6.4 milestone to add `wp_trigger_error()`. I agree with you @azaozz that it can be modified in the future when it's time to implement input (function/method parameter) validation.

- For this PR as a compromise, I'll switch `trigger_error()` to `_doing_it_wrong()` with a comment to replace it with `wp_trigger_error()`.

hellofromtonya replied:
Done in commit https://github.com/WordPress/wordpress-develop/pull/5051/commits/e695f6f06de51f304b3686694d0197e2f832813b ✅

felixarntz commented on src/wp-includes/fonts/fonts.php:
```diff
@@ -0,0 +1,63 @@
+ *             @type string          $font_style              Optional. The font-style property. Default 'normal'.
+ *             @type string          $font-weight             Optional. The font-weight property. Default '400'.
+ *             @type string          $font-display            Optional. The font-display property. Default 'fallback'.
+ *             The following font-face properties are optional and will not be generated if not provided.
```
See my comment above, this line should be removed.

Additionally, to ease maintenance, I would advise to only keep the detailed list of all fields in one place (probably this function, since it's the primarily recommended one).

hellofromtonya replied:
Fixed https://github.com/WordPress/wordpress-develop/pull/5051/commits/31a3508a839cfa0427c2c471bc8513b0693e8ed9  ✅

felixarntz commented on src/wp-includes/fonts/font-face/class-wp-font-face.php:
```diff
@@ -0,0 +1,439 @@
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array[][] $fonts {
+	 *     Optional. The font-families and their font variations. Default empty array.
+	 *
+	 *     @type string $font-family => array[] $variations {
+	 *         Optional. An associated array of font variations for this font-family.
+	 *         Each variation has the following structure.
+	 *
+	 *         @type array $font_variation {
+	 *             @type string          $font-family             The font-family property.
+	 *             @type string|string[] $src                     The URL(s) to each resource containing the font data.
+	 *             @type string          $font_style              Optional. The font-style property. Default 'normal'.
+	 *             @type string          $font-weight             Optional. The font-weight property. Default '400'.
+	 *             @type string          $font-display            Optional. The font-display property. Default 'fallback'.
+	 *             The following font-face properties are optional and will not be generated if not provided.
+	 *             @type string          $ascent-override         Optional. The ascent-override property.
+	 *             @type string          $descent-override        Optional. The descent-override property.
+	 *             @type string          $font-stretch            Optional. The font-stretch property.
+	 *             @type string          $font-variant            Optional. The font-variant property.
+	 *             @type string          $font-feature-settings   Optional. The font-feature-settings property.
+	 *             @type string          $font-variation-settings Optional. The font-variation-settings property.
+	 *             @type string          $line-gap-override       Optional. The line-gap-override property.
+	 *             @type string          $size-adjust             Optional. The size-adjust property.
+	 *             @type string          $unicode-range           Optional. The unicode-range property.
+	 *         }
+	 *     }
```
See my other comment, I think for maintenance purposes and to prevent things getting out of sync, we should keep these docs in a single place, either here or in `wp_print_font_faces()`. The respective other function should then refer to that, for example like this:
```suggestion
	 * @param array[][] $fonts Optional. The font-families and their font variations.
	 *                         See {@see wp_print_font_faces()} for the supported fields.
	 *                         Default empty array.
```

hellofromtonya replied:
Thank you @felixarntz. Great suggestion! Committed your suggestion in commit https://github.com/WordPress/wordpress-develop/pull/5051/commits/f6c42668cf426456a5d5e36e1ad5d15636dc523e ✅

felixarntz reviewed (changes requested):
@hellofromtonya Mostly the PR looks solid to me. I have a few points of feedback below.

One broader nit-pick, I'm not sure about the file organization of the new files, I think all the nested folders are unnecessary and don't fit well with the way core typically organizes files. I would instead advise the following structure:
* `wp-includes/fonts` directory for the new classes (no additional subdirectories)
* `wp-includes/fonts.php` file for the global function(s)

hellofromtonya commented:
>One broader nit-pick, I'm not sure about the file organization of the new files, I think all the nested folders are unnecessary and don't fit well with the way core typically organizes files. I would instead advise the following structure:
>
>wp-includes/fonts directory for the new classes (no additional subdirectories)
>wp-includes/fonts.php file for the global function(s)

Hello @felixarntz 👋  Thank you for your review!

The thinking around the file organization is to group the different fonts management components in sub-directories, which as of today is Font Face and Font Library. There's a mix of approaches in Core, though predominately there is as you note a tendency to group all like files in the "group" folder, rather than sub directories.

I'm okay with either approach. When Font Library lands, it would also go into the `wp-includes/fonts/` directory too.

hellofromtonya committed:
Reference @fonts param definition

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

hellofromtonya committed:
Remove fonts param explanation line

hellofromtonya committed:
Remove "wp_font_face_property_defaults" filter

hellofromtonya committed:
Relocate into fonts/

hellofromtonya commented:
@felixarntz commit https://github.com/WordPress/wordpress-develop/pull/5051/commits/e2db981bc075cc27769f202fd2133ce1388c0de7 relocates the `wp-includes/fonts/font-face` files into its root `wp-includes/fonts/` per your suggestion ✅

felixarntz commented:
@hellofromtonya Thank you for the updates, also for updating the file/folder structure. What about moving the functions file `fonts.php` directly into `wp-includes`, i.e. `wp-includes/fonts.php`? I think that ties in better with how Core organizes files, it is common to have general global function files in there and then specific classes for a feature in a directory. For example, that is similarly handled for the REST API, which uses `wp-includes/rest-api.php` for functions and the `wp-includes/rest-api` directory for classes.

costdev commented on src/wp-includes/fonts/class-wp-font-face.php:
```diff
@@ -0,0 +1,400 @@
+		$css = '';
+
+		// Wrap font-family in quotes if it contains spaces
+		// and is not already wrapped in quotes.
```
```suggestion
		/*
		 * Wrap font-family in quotes if it contains spaces
		 * and is not already wrapped in quotes.
		 */
```
See [the handbook entry](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#5-2-multi-line-comments) for how Core formats multi-line comments.

hellofromtonya replied:
Thank you @costdev! Done in https://github.com/WordPress/wordpress-develop/pull/5051/commits/4a1d4c2b55ee54cb8563ad06ed9237ce5f73ab6a ✅

costdev commented on src/wp-includes/fonts/class-wp-font-face.php:
```diff
@@ -0,0 +1,400 @@
+		}
+
+		// Validate the 'src' property.
+		if ( ! empty( $font_face['src'] ) ) {
```
The previous lines ensure that this can't be empty by now. Suggest removing this condition.

hellofromtonya replied:
Thank you @costdev! Done in https://github.com/WordPress/wordpress-develop/pull/5051/commits/4a1d4c2b55ee54cb8563ad06ed9237ce5f73ab6a ✅

costdev commented on src/wp-includes/fonts/class-wp-font-face.php:
```diff
@@ -0,0 +1,400 @@
+
+		// Check the font-display.
+		if ( ! in_array( $font_face['font-display'], $this->valid_font_display, true ) ) {
+			$font_face['font-display'] = $this->font_face_property_defaults['font-display'];
```
It might be helpful to trigger an error or `_doing_it_wrong()` here to indicate an invalid `font-display` value.

hellofromtonya replied:
Hmm, it's not a doing it wrong, as an invalid value will use the default instead and the processing continues (unlike the other validation above where validation fails and it bails out).

Looking at Core, other loosely similar checks do not inform the developer either. That makes me wonder .. Should Core alert that it is using a default value if the given value fails its validation? Hmm, interesting 🤔. It might be helpful to developers to help them debug.

If this were to throw a notice, likely should wrap that notice in `empty( trim () )` to avoid the notice on an empty string value.

As this isn't a `_doing_it_wrong()` and `wp_trigger_error()` does not yet exist in Core, this helper message could be added later once `wp_trigger_error()` is introduced. It's definitely worth a good think and discussion.

costdev commented on src/wp-includes/fonts/class-wp-font-face.php:
```diff
@@ -0,0 +1,400 @@
+		// Remove invalid properties.
+		foreach ( $font_face as $prop => $value ) {
+			if ( ! in_array( $prop, $this->valid_font_face_properties, true ) ) {
+				unset( $font_face[ $prop ] );
```
See above regarding indicating invalid properties/values.

hellofromtonya replied:
Thank you @costdev! Done in https://github.com/WordPress/wordpress-develop/pull/5051/commits/4a1d4c2b55ee54cb8563ad06ed9237ce5f73ab6a ✅

costdev commented on src/wp-includes/fonts/class-wp-font-face.php:
```diff
@@ -0,0 +1,400 @@
+			return;
+		}
+
+		printf(
+			$this->get_style_element(),
+			$this->get_css( $fonts )
+		);
```
As `$fonts` can come from anywhere, we should escape here to avoid evils slipping through from the `::get_css()` callstack - happy to discuss details in DM.

hellofromtonya replied:
Hmm, as the CSS string is within the style element, the nefarious code would need to inject `</style>` before its evil code. Alrighty, then tags need to be stripped from the CSS string to ensure that doesn't happen. `wp_strip_all_tags()` should be sufficient. 

BTW: This is the approach used in TT0 and TT1 for their CSS auto-generator.

For example, here's TT1:
```php
function twenty_twenty_one_generate_css( $selector, $style, $value, $prefix = '', $suffix = '', $display = true ) {

	// Bail early if there is no $selector elements or properties and $value.
	if ( ! $value || ! $selector ) {
		return '';
	}

	$css = sprintf( '%s { %s: %s; }', $selector, $style, $prefix . $value . $suffix );

	if ( $display ) {
		/*
		 * Note to reviewers: $css contains auto-generated CSS.
		 * It is included inside <style> tags and can only be interpreted as CSS on the browser.
		 * Using wp_strip_all_tags() here is sufficient escaping to avoid
		 * malicious attempts to close </style> and open a <script>.
		 */
		echo wp_strip_all_tags( $css ); // phpcs:ignore WordPress.Security.EscapeOutput
	}
	return $css;
}
```

hellofromtonya replied:
Added a test for it.

Using this data:
```php
$fonts = array(
	'Source Serif Pro' => array(
		array(
			'src'          => array( 'http://example.com/assets/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2' ),
			'font-family'  => 'Source Serif Pro',
			'font-style'   => 'normal',
			'font-weight'  => '200 900',
			'font-stretch' => '</style><script>console.log("Hello")</script><style>',
		),
	),
);
```


Before `wp_strip_all_tags( $css )`:
```html
<style id='wp-fonts-local' type='text/css'>
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:url('http://example.com/assets/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');font-stretch:</style><script>console.log("Hello")</script><style>;}
</style>
```

After `wp_strip_all_tags( $css )`:

```html
<style id='wp-fonts-local' type='text/css'>
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:url('http://example.com/assets/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');font-stretch:;}
</style>
```

Excellent catch @costdev 👏

hellofromtonya replied:
Added `wp_strip_all_tags( $css )` in the source code via commit https://github.com/WordPress/wordpress-develop/pull/5051/commits/e90b21d7b1f6a6857710c50fc69b2a1f3672175d ✅ 

Added a test for it in commit https://github.com/WordPress/wordpress-develop/pull/5051/commits/b4bcd076314387670fb32559072d092782fb217f ✅

costdev commented on src/wp-includes/fonts/class-wp-font-face-resolver.php:
```diff
@@ -0,0 +1,154 @@
+		$settings = wp_get_global_settings();
+
+		// Bail out early if there are no font settings.
+		if ( empty( $settings['typography'] ) || empty( $settings['typography']['fontFamilies'] ) ) {
```
Shouldn't the second condition cover the first?

hellofromtonya replied:
Actually, the second condition is all that's needed as it'll catch both conditions https://3v4l.org/b6UYn.

hellofromtonya replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5051/commits/4a1d4c2b55ee54cb8563ad06ed9237ce5f73ab6a ✅

costdev commented on src/wp-includes/fonts/class-wp-font-face.php:
```diff
@@ -0,0 +1,400 @@
+	 * @since 6.4.0
+	 *
+	 * @param array $font_face Font face properties to validate.
+	 * @return false|array Validated font-face on success. Else, false.
```
```suggestion
	 * @return array|false Validated font-face on success. Otherwise false.
```
- Swapped return types. You pessimist! :joy:
- "Otherwise false" is Core's usual phrasing.
- Also, since this validates both properties and values, and returns the validated font face on success rather than valid properties, does it make more sense to call this `validate_font_face()`?

hellofromtonya replied:
Good catch on the swapped return types. Whoopsie. 

>"Otherwise false" is Core's usual phrasing.

Hmm, there's a mixture of phrasing in Core:
- `. Otherwise false`
- [the success]`, false on failure.`
- , otherwise false.'
- `, or false on failure.`

I think I'll change it to the last one.

hellofromtonya replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5051/commits/4a1d4c2b55ee54cb8563ad06ed9237ce5f73ab6a ✅

hellofromtonya replied:
>Also, since this validates both properties and values, and returns the validated font face on success rather than valid properties, does it make more sense to call this validate_font_face()?

I think that method name would be confusing due to:

* The font-face array contains a collection of `@font-face` property / value pairings that relate directly to the CSS. 
*  The method above it called `validate_fonts()`.

`properties` here is a verbose attempt to infer property / value pairs.

Other more accurate names could be:

* `validate_font_face_property_value_pairs()`
* `validate_font_face_declarations()` <= [the proper name for a property / value pair](https://developer.mozilla.org/en-US/docs/Web/CSS/Syntax#css_declarations)

@costdev Do you find `properties` to be confusing? Is either of the above more understandable?

costdev replied:
Thanks for the clarification!

I think `properties` while verbose was also a bit vague as the code validates properties and values.

`validate_font_face_declarations()` makes more sense to me as it encompasses both, so there's no surprises when seeing the internal code.

hellofromtonya replied:
Done in commit https://github.com/WordPress/wordpress-develop/pull/5051/commits/c8e4c03546418aae76f62ba1dc07dd2193b847e7 ✅ 

Also updated the DocBlock to ensure "declaration" is well understood.

costdev commented on src/wp-includes/fonts/class-wp-font-face.php:
```diff
@@ -0,0 +1,400 @@
+		}
+
+		// Remove invalid properties.
+		foreach ( $font_face as $prop => $value ) {
```
Nit: IMO, there's no need to abbreviate to "prop" here. "property" and "properties" are used in documentation, as well as property and method names throughout.

Note the [coding standards on naming conventions](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#naming-conventions):
> Don’t abbreviate variable names unnecessarily; let the code be unambiguous and self-documenting.

hellofromtonya replied:
Thank you @costdev! Done in https://github.com/WordPress/wordpress-develop/pull/5051/commits/4a1d4c2b55ee54cb8563ad06ed9237ce5f73ab6a ✅

costdev commented on src/wp-includes/fonts/class-wp-font-face.php:
```diff
@@ -0,0 +1,400 @@
+	}
+
+	/**
+	 * Gets the `<style>` element for wrapping the `@font-face` CSS.
```
```suggestion
	 * Gets the style element for wrapping the font-face CSS.
```
The documentation standards state:
> No HTML markup or Markdown of any kind should be used in the summary. If the text refers to an HTML element or tag, then it should be written as “image tag” or “img” element, not “< img >“. (spaces added by me)
> [Reference](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#summary)

:1234: Applies elsewhere.

hellofromtonya replied:
Thank you @costdev! Done in https://github.com/WordPress/wordpress-develop/pull/5051/commits/4a1d4c2b55ee54cb8563ad06ed9237ce5f73ab6a ✅

costdev reviewed (changes requested):
Nothing major, just a few thoughts/queries in this review.

I'll get this tested today. :slightly_smiling_face:

anton-vlasenko commented on src/wp-includes/deprecated.php:
```diff
@@ -5367,3 +5367,506 @@ function block_core_navigation_submenu_build_css_colors( $context, $attributes,
+
+				// Combine variations with settings. Remove duplicates.
+				$settings['typography']['fontFamilies']['theme'] = array_merge( $settings['typography']['fontFamilies']['theme'], $variation['settings']['typography']['fontFamilies']['theme'] );
+				$settings['typography']['fontFamilies']          = array_unique( $settings['typography']['fontFamilies'] );
```
`$settings['typography']['fontFamilies']` is an array of arrays, so calling `array_unique()` on it might not work as expected.
Should `array_unique()` be called on the `['theme']` key specifically?
```suggestion
				$settings['typography']['fontFamilies']['theme'] = array_unique( $settings['typography']['fontFamilies']['theme'] );
```

hellofromtonya replied:
@anton-vlasenko Thank you :) 

This code is being deprecated and completely replaced by Font Face. It seems highly unlikely that an extender would be using it. Thus, I think it's best not to modify its code.

anton-vlasenko commented on src/wp-includes/deprecated.php:
```diff
@@ -5367,3 +5367,506 @@ function block_core_navigation_submenu_build_css_colors( $context, $attributes,
+		$src         = array();
+		$src_ordered = array();
+
+		foreach ( $webfont['src'] as $url ) {
+			// Add data URIs first.
+			if ( str_starts_with( trim( $url ), 'data:' ) ) {
+				$src_ordered[] = array(
+						'url'    => $url,
+						'format' => 'data',
+				);
+				continue;
+			}
+			$format         = pathinfo( $url, PATHINFO_EXTENSION );
+			$src[ $format ] = $url;
+		}
+
+		// Add woff2.
+		if ( ! empty( $src['woff2'] ) ) {
+			$src_ordered[] = array(
+					'url'    => sanitize_url( $src['woff2'] ),
+					'format' => 'woff2',
+			);
+		}
```
I'm wondering if the woff2 format should come first, as woff2 fonts typically have a smaller size compared to Data URI, being about 30% less.
Applies elsewhere.

hellofromtonya replied:
@anton-vlasenko Thank you :) 

This code is being deprecated and completely replaced by Font Face. It seems highly unlikely that an extender would be using it. Thus, I think it's best not to modify its code.

anton-vlasenko commented on src/wp-includes/fonts/class-wp-font-face.php:
```diff
@@ -0,0 +1,400 @@
+	 * @param array[][] $fonts Optional. The font-families and their font variations.
+	 *                         See {@see wp_print_font_faces()} for the supported fields.
+	 *                         Default empty array.
+	 * }
```
Nitpick: this curly brace doesn't seem to be necessary here.
```suggestion
	 *
```

hellofromtonya replied:
Thank you @anton-vlasenko! Done in https://github.com/WordPress/wordpress-develop/pull/5051/commits/4a1d4c2b55ee54cb8563ad06ed9237ce5f73ab6a ✅

anton-vlasenko replied:
Thank you!

anton-vlasenko reviewed (commented):
Please find my notes below.
Thank you.

hellofromtonya committed:
Move fonts.php to wp-includes/

hellofromtonya committed:
Replace trigger_error()

1. Uses _doing_it_wrong() until wp_trigger_error() is available.
2. Adds comment to each with `@todo` annotation as a reminder to replace these later.

felixarntz reviewed (approved):
@hellofromtonya This looks solid to me now. Other than the oustanding minor quirks outlined by other reviewers, I don't have any blockers on this one. Excited to see it land!

It would be great to do a performance comparison for how this does compared to the old private code, just to ensure there's no unexpected regression.

hellofromtonya committed:
Code reviews from Anton and Colin

hellofromtonya committed:
Escape CSS for all tags

hellofromtonya committed:
CSS: strip all tags before printing

hellofromtonya committed:
Rename properties to declarations

costdev reviewed (approved):
Thanks for the updates @hellofromtonya! This looks good to me!

hellofromtonya committed:
Run tests in separate processes

hellofromtonya commented:
Committed via https://core.trac.wordpress.org/changeset/53282. Thank you everyone for your contributions!
